lib/rules/no-restricted-imports.js

```diff
@@ -234,20 +245,53 @@ module.exports = {
         /**
          * Report a restricted path specifically for patterns.
          * @param {node} node representing the restricted path reference
-         * @param {Object} group contains a Ignore instance for paths, and the customMessage to show if it fails
+         * @param {Object} group contains an Ignore instance for paths, the customMessage to show on failure,
+         * and any restricted import names that have been specified in the config
+         * @param {Map<string,Object[]>} importNames Map of import names that are being imported
          * @returns {void}
          * @private
          */
-        function reportPathForPatterns(node, group) {
+        function reportPathForPatterns(node, group, importNames) {
             const importSource = node.source.value.trim();
 
-            context.report({
-                node,
-                messageId: group.customMessage ? "patternWithCustomMessage" : "patterns",
-                data: {
-                    importSource,
-                    customMessage: group.customMessage
+            const customMessage = group.customMessage;
+            const restrictedImportNames = group.importNames || [];
+
+            /*
+             * If we are not restricting to any specific import names and just the pattern itself,
+             * report the error and move on
+             */
+            if (restrictedImportNames.length < 1 || importNames.has("*")) {
```

Comment:
The message for `*` would be misleading, as it would suggest that the whole module is restricted:

```js
/* eslint no-restricted-imports: ["error", {
    patterns: [{
        group: ["mod"],
        importNames: ["foo"]
    }]
}]*/

import * as b from "mod"; // 'mod' import is restricted from being used by a pattern.

```

We should handle `*` here in the same way as it is [handled for paths](https://github.com/eslint/eslint/blob/b6aee9591ecc2e2f5738ab8bef20faac1e05b5c3/lib/rules/no-restricted-imports.js#L189), and use messages like [`everything` and `everythingWithCustomMessage`](https://github.com/eslint/eslint/blob/b6aee9591ecc2e2f5738ab8bef20faac1e05b5c3/lib/rules/no-restricted-imports.js#L105-L107) with a note that the restriction is by a pattern (we'll need two new messageIds).

Reply:
It's a different case because there are no `importNames` specified, so the entire module is restricted.

When `importNames` are specified, we are disallowing `*` because those names can be accessed through it.

You can check how this works for paths:

```js
/* eslint no-restricted-imports: ["error", {
    paths: [
        {
            name: "mod1"            
        },
        {
            name: "mod2",
            importNames: ["foo"]
        }
    ]
}]*/

import * as a from "mod1"; // 'mod1' import is restricted from being used.

import * as b from "mod2"; // * import is invalid because 'foo' from 'mod2' is restricted.
```

Reply:
Ah, okay - I see. That makes more sense, thank you! I should be able to make that change tonight.